runtime-sdk/modules/evm/src/lib.rs

```diff
@@ -356,9 +522,103 @@ impl module::MigrationHandler for Module {
     }
 }
 
-impl module::AuthHandler for Module {}
+impl<Accounts: modules::accounts::API> module::AuthHandler for Module<Accounts> {
+    fn authenticate_tx<C: Context>(ctx: &mut C, tx: &Transaction) -> Result<(), CoreError> {
+        // We're only interested in transactions that can be paid with tokens
+        // from the corresponding EVM account.
+        if !(tx.call.method == "evm.Create"
+            || tx.call.method == "evm.Call"
+            || tx.call.method == "evm.Withdraw")
+        {
+            return Err(CoreError::AuthenticateTxNotHandled);
+        }
+
+        let params = Self::params(ctx.runtime_state());
+        let den = params.token_denomination;
+        let min_gas_price = params.min_gas_price;
+
+        let evm_acct_addr = Self::derive_caller_from_tx_auth_info(&tx.auth_info)
+            .map_err(|e| CoreError::MalformedTransaction(anyhow::Error::new(e)))?;
+
+        // We're going to withdraw the maximum possible fees for the EVM
+        // transaction -- the difference between the max and the actual
+        // fees will be returned in do_evm above, once the execution is
+        // finished and we have calculated the actual fees.
+        // In addition, we will also withdraw Oasis fees for this tx from
+        // the same EVM account.
+
+        let oasis_fees = tx.auth_info.fee.amount.amount();
+        let evm_fees: u128 = match tx.call.method.as_str() {
+            "evm.Withdraw" => 0,
+            "evm.Call" => {
+                let call: CallTx = cbor::from_value(tx.call.body.clone()).map_err(|_| {
+                    CoreError::MalformedTransaction(format_err!(
+                        "evm.Call call body doesn't decode into CallTx"
+                    ))
+                })?;
+                // Make sure that gas_price >= min_gas_price.
+                if call.gas_price < min_gas_price {
+                    return Err(CoreError::MalformedTransaction(anyhow::Error::new(
+                        Error::GasPriceTooLow,
+                    )));
+                }
+                let evm_fee = call
+                    .gas_price
+                    .checked_mul(U256::from(call.gas_limit))
+                    .ok_or(CoreError::GasOverflow)?
+                    .checked_add(call.value)
```

Comment:
Value should probably be separate from the gas fee as that is only moved in case the transaction is successful? Also below.